### DISCUSSION TITLE
如何对多种配置进行并行编译

### DISCUSSION THREAD

**Discussion by MasterBalll** [xmake-io/xmake#3338] [Q&A] [answered]
有这样一种需求，就是要在 windows 与 cross 下对多种配置进行编译、运行（具体来说，是对几个嵌入式平台的仿真以及对这若干目标平台程序进行验证）。
顺序执行会这么做，比如，
```shell
xmake f -p windows -a x86 -m release --opt1=A
xmake
xmake r
xmake f -p windows -a x86 -m release --opt1=B
xmake
xmake r
xmake f -p cross -a xtensa -m release --opt1=A   # 这里使用自定义的嵌入式工具链
xmake
xmake r
xmake f -p cross -a xtensa -m release --opt1=B   # 同上
xmake
xmake r
```
但这样会比较浪费时间，特别是切换配置时似乎要重新检查编译、链接选项等，想问下有无好办法？

**Comment by waruqi** [comment, marked as answer]
https://xmake.io/#/zh-cn/plugin/builtin_plugins?id=%e5%ae%8f%e8%ae%b0%e5%bd%95%e5%92%8c%e5%9b%9e%e6%94%be

> 特别是切换配置时似乎要重新检查编译、链接选项等，想问下有无好办法？

不加 -c，默认 build 分 plat/mode/arch 存储，切平台和arch/mode 不会重新检测，切配置当然要重新检查的，否则不保证编译正确。。

**Reply by MasterBalll** [reply]
第一个小问题：
`set_targetdir("$(buildir)/$(plat)/$(arch)/$(opt1)/$(mode)")` 
这样设置，切配置也要重新检查吗？想把 opt1 这样的选项当作一个和 plat/mode/arch 同级别的东西。

第二个小问题：
我倒是写了个简单的 task 来稍微简化下配置，也类似于 macro 了，不过这样还是顺序执行，不知道有没有什么标准的办法能做到并行。
之前尝试过用循环来定义 target，这样不同配置对应不同名字的 target，可以并行编译；但如果配置多起来，target 数量就非常多，有人喜欢用生成的 vs 工程来开发，这样一来就很碍眼，一个维护者一般只需要操纵一部分配置就好了。
就很纠结该怎么做。

**Reply by waruqi** [reply]
> 这样设置，切配置也要重新检查吗？想把 opt1 这样的选项当作一个和 plat/mode/arch 同级别的东西。

切 -p/-a/-m 不会，切其他配置会。

> 我倒是写了个简单的 task 来稍微简化下配置，也类似于 macro 了，不过这样还是顺序执行，不知道有没有什么标准的办法能做到并行。

没必要也不支持，每次配置，所有源码都已经并行编译 跑满 cpu core了。。你 切配置 再搞成 并行，只会打乱缓存，不会有任何加速。。
